src/bin/ipa_bench/gen_events.rs

```diff
@@ -16,35 +16,84 @@ use std::time::Duration;
 const RECORD_SEPARATOR: u8 = 30;
 
 const DAYS_IN_EPOCH: u64 = 7;
-type MatchKey = Vec<u64>;
+type MatchKey = u64;
 type Epoch = u8;
+type Number = u32;
 
 #[derive(Clone)]
 #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
 pub struct EventBase {
-    // For this tool, we'll fix the length of a matchkey to u64
-    pub matchkeys: MatchKey,
+    /// Secret shared and then encrypted match keys.
+    pub matchkeys: Vec<MatchKey>,
+
+    /// The epoch which this event is generated. Using an 8-bit value = 256 epochs > 4 years (assuming 1 epoch = 1 week).
+    /// This field is in the clear.
     pub epoch: Epoch,
-    pub timestamp: u32,
+
+    /// An offset in seconds into a given epoch. The clear is u32 (< 2^20 seconds), then encrypted and secret shared.
+    pub timestamp: Number,
+
+    /// 0 (false) if this is a source event. 1 (true) otherwise.
+    pub is_trigger: bool,
```

Comment:
Don't you want to use an enum for this?  That is, something like:

```rust
pub enum Event {
  Source {
    event: EventBase,
    breakdown_key: Number,
  },
  Trigger {
    event: EventBase,
    value: Number,
  },
}
```
I realize that you want to have the binary value of source and trigger events put the `breakdown_key` and `value` fields in different positions, but for that you will want to implement `Serialize` and `Deserialize` manually anyway.

Reply:
Makes sense. Let me work on implementing Serialize/Deserialize manually in a separate diff. I think the work would be large enough that it should be reviewed separately. Also, I'll be on the leave for a month so I want this PR to land before I'm gone if possible.